File: python/linked-list/linked_list.py
class Node:
    def __init__(self, value, succeeding=None, previous=None):
        self.value = value
        self.succeeding = succeeding
        self.previous = previous
class LinkedList:
    def __init__(self):
        head_guard_node = Node(None)
        tail_guard_node = Node(None)
        head_guard_node.succeeding = tail_guard_node
        tail_guard_node.previous = head_guard_node
        self.head = head_guard_node
        self.tail = tail_guard_node
        self.length = 0
    # insert value at back
    def push(self, value):
        self.insert(self.tail.previous, self.tail, value)
    # remove value at back
    def pop(self):
        return self.remove(self.tail.previous)
    # insert value at front
    def unshift(self, value):
        self.insert(self.head, self.head.succeeding, value)
    # remove value at front
    def shift(self):
        return self.remove(self.head.succeeding)
    def __len__(self):
        return self.length
    def delete(self, value):
        if self.length == 0:
            raise ValueError('Value not found')
        curr_node = self.head
        while curr_node is not None:
            if curr_node.value == value:
                self.remove(curr_node)
                return
            curr_node = curr_node.succeeding
        raise ValueError('Value not found')
    def insert(self, previous, succeeding, value):
        new_node = Node(value, succeeding, previous)
        succeeding.previous = new_node
        previous.succeeding = new_node
        self.length += 1
    def remove(self, node):
        if self.length == 0:
            raise IndexError('List is empty')
        node.previous.succeeding = node.succeeding
        node.succeeding.previous = node.previous
        self.length -= 1
        return node.value
        

          


            


          

File: javascript/linked-list/linked-list.js
//
// This is only a SKELETON file for the 'Linked List' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

class Node {
  constructor(value) {
    this.value = value; // Store the value of the node
    this.next = null; // Initialize the next pointer to null
  }
}

export class LinkedList {
  constructor() {
    this.head = null; // Initialize the linked list with an empty head
  }

  // Push a new element to the end of the linked list
  push(value) {
    const newNode = new Node(value); // Create a new node with the given value
    if (!this.head) {
      this.head = newNode; // If the list is empty, set the new node as the head
    } else {
      let current = this.head;
      while (current.next) {
        current = current.next; // Traverse to the end of the list
      }
      current.next = newNode; // Set the new node as the next element of the current node
    }
  }

  // Remove and return the last element from the linked list
  pop() {
    if (!this.head) {
      return undefined; // If the list is empty, return undefined
    }
    if (!this.head.next) {
      const value = this.head.value;
      this.head = null; // If there is only one element, remove it
      return value;
    }
    let current = this.head;
    let previous = null;
    while (current.next) {
      previous = current;
      current = current.next; // Traverse to the end of the list while tracking the previous node
    }
    previous.next = null; // Set the next of the previous node to null
    return current.value; // Return the value of the removed node
  }

  // Remove and return the first element from the linked list
  shift() {
    if (!this.head) {
      return undefined; // If the list is empty, return undefined
    }
    const value = this.head.value;
    this.head = this.head.next; // Update the head to the next node
    return value;
  }

  // Add a new element to the beginning of the linked list
  unshift(value) {
    const newNode = new Node(value); // Create a new node with the given value
    newNode.next = this.head; // Set the new node's next to the current head
    this.head = newNode; // Update the head to the new node
  }

  // Delete the first occurrence of a value from the linked list
  delete(value) {
    if (!this.head) {
      return; // If the list is empty, do nothing
    }
    if (this.head.value === value) {
      this.head = this.head.next; // If the head has the value, remove the head
      return;
    }
    let current = this.head;
    let previous = null;
    while (current && current.value !== value) {
      previous = current;
      current = current.next; // Traverse through the list to find the value
    }
    if (current) {
      previous.next = current.next; // Remove the node with the value
    }
  }

  // Count and return the number of elements in the linked list
  count() {
    let count = 0;
    let current = this.head;
    while (current) {
      count++;
      current = current.next; // Traverse through the list and count the nodes
    }
    return count;
  }
}
